python/tank/platform/interactive_login.py

```diff
@@ -0,0 +1,208 @@
+# Copyright (c) 2015 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+UI and console based login for Toolkit.
+"""
+
+from getpass import getpass
+import logging
+
+from tank_vendor.shotgun_api3 import Shotgun, AuthenticationFault
+from tank.util.session import cache_session_data, get_cached_login_info, delete_session_data, is_session_token_cached
+from tank.util.login import get_login_name
+from tank.util.shotgun import get_associated_sg_config_data, get_associated_sg_base_url
+
+# Configure logging
+logger = logging.getLogger("sgtk-interactive_login")
+# logger.setLevel(logging.DEBUG)
+logger.addHandler(logging.StreamHandler())
+
+
+def _get_session_token(hostname, login, password, http_proxy):
+    """
+    Retrieves a session token for the given credentials.
+    :param hostname: The host to connect to.
+    :param login: The user to get a session for.
+    :param password: Password for the user.
+    :param http_proxy: Proxy to use. Can be None.
+    :returns: If the credentials were valid, returns a session token, otherwise returns None.
+    """
+    try:
+        # Create the instance...
+        sg = Shotgun(
+            hostname.encode("utf-8"),
+            login=login.encode("utf-8"),
+            password=password.encode("utf-8"),
+            http_proxy=http_proxy.encode("utf-8") if http_proxy else None
+        )
+        # .. and generate the session token. If it throws, we have invalid credentials.
+        return sg.get_session_token()
+    except AuthenticationFault:
+        print "Authentication failed."
```

Comment:
Yes, it will be refactored in the future. We could then have a authentication_fault_handler functor passed in that would print on the console or pop a message box.

My code is partly already refactored, because at one point I had some UI code in there. When we decided to scope down the ticket to only do console stuff, I removed the few lines or Qt related code I had. However, some method that had _console_ in the name stuck around.

From: Manne Öhrström <notifications@github.com<mailto:notifications@github.com>>
Reply-To: shotgunsoftware/tk-core <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Tuesday, February 17, 2015 at 9:21 AM
To: shotgunsoftware/tk-core <tk-core@noreply.github.com<mailto:tk-core@noreply.github.com>>
Cc: JF Boismenu <jean.francois.boismenu@autodesk.com<mailto:jean.francois.boismenu@autodesk.com>>
Subject: Re: [tk-core] For #28030, [security] User/password sign in on tank command with session caching (#129)


In python/tank/platform/interactive_login.py<https://github.com/shotgunsoftware/tk-core/pull/129#discussion_r24818002>:

> +    :param password: Password for the user.
> +    :param http_proxy: Proxy to use. Can be None.
> +    :returns: If the credentials were valid, returns a session token, otherwise returns None.
> +    """
> +    try:
> +        # Create the instance...
> +        sg = Shotgun(
> +            hostname.encode("utf-8"),
> +            login=login.encode("utf-8"),
> +            password=password.encode("utf-8"),
> +            http_proxy=http_proxy.encode("utf-8") if http_proxy else None
> +        )
> +        # .. and generate the session token. If it throws, we have invalid credentials.
> +        return sg.get_session_token()
> +    except AuthenticationFault:
> +        print "Authentication failed."


gotcha! Do you think this method will be use by _do_ui_based_login later on? I just thought that maybe later it may be called from the ui side of things too.... or maybe it can be refactored at that point? The name _get_session_token seemed to imply that it is a general method and not intended for console based use!

—
Reply to this email directly or view it on GitHub<https://github.com/shotgunsoftware/tk-core/pull/129/files#r24818002>.